src/Field.php

```diff
@@ -475,49 +341,66 @@ public function useAlias()
 
     // }}}
 
-    // {{{ Handy methods used by UI
+    // {{{ Handy methods used by UI and in other places
 
     /**
      * Returns if field should be editable in UI.
      *
      * @return bool
      */
-    public function isEditable()
+    public function isEditable() : bool
     {
-        return isset($this->ui['editable']) ? $this->ui['editable']
-                : (($this->read_only || $this->never_persist) ? false
-                    : !$this->system);
+        return $this->ui['editable'] ?? (($this->read_only || $this->never_persist) ? false : !$this->system);
     }
 
     /**
      * Returns if field should be visible in UI.
      *
      * @return bool
      */
-    public function isVisible()
+    public function isVisible() : bool
     {
-        return isset($this->ui['visible']) ? $this->ui['visible'] : !$this->system;
+        return $this->ui['visible'] ?? !$this->system;
     }
 
     /**
      * Returns if field should be hidden in UI.
      *
      * @return bool
      */
-    public function isHidden()
+    public function isHidden() : bool
     {
-        return isset($this->ui['hidden']) ? $this->ui['hidden'] : false;
+        return $this->ui['hidden'] ?? false;
+    }
+
+    /**
+     * Returns true if field allows NULL values.
+     *
+     * @return bool
+     */
+    public function canBeNull() : bool
+    {
+        return $this->mandatory === false;
+    }
+
+    /**
+     * Returns true if field allows EMPTY values like empty string and NULL.
+     *
+     * @return bool
+     */
+    public function canBeEmpty() : bool
+    {
+        return $this->mandatory === false && $this->required === false;
```

Comment:
return !$this->['ui']['required'] ?? true;

Reply:
@romaninsh '$this->['ui']['required']' <--- i lost some meeting but why check for UI required in place of required ?